diff -ruN gcc-releases-gcc-13.4.0/gcc/c-family/c-attribs.cc ../_temp/gcc.work/gcc/c-family/c-attribs.cc
--- gcc-releases-gcc-13.4.0/gcc/c-family/c-attribs.cc	2025-06-05 18:03:58.000000000 +0200
+++ ../_temp/gcc.work/gcc/c-family/c-attribs.cc	2026-03-17 20:59:22.127555336 +0100
@@ -137,10 +137,9 @@
 static tree handle_nonstring_attribute (tree *, tree, tree, int, bool *);
 static tree handle_nothrow_attribute (tree *, tree, tree, int, bool *);
 static tree handle_cleanup_attribute (tree *, tree, tree, int, bool *);
-static tree handle_warn_unused_result_attribute (tree *, tree, tree, int,
-						 bool *);
+static tree handle_warn_unused_result_attribute (tree *, tree, tree, int, bool *);
 static tree handle_access_attribute (tree *, tree, tree, int, bool *);
-
+static tree handle_libcall_attribute (tree *, tree, tree, int, bool *);
 static tree handle_sentinel_attribute (tree *, tree, tree, int, bool *);
 static tree handle_type_generic_attribute (tree *, tree, tree, int, bool *);
 static tree handle_alloc_size_attribute (tree *, tree, tree, int, bool *);
@@ -443,6 +442,11 @@
   { "warn_unused_result",     0, 0, false, true, true, false,
 			      handle_warn_unused_result_attribute,
 	                      attr_warn_unused_result_exclusions },
+  { "libcall",                0, 0, false, true, true, false,
+                              handle_libcall_attribute, NULL },
+  /* Similiar to libcall but doesn't imply linearvarargs. Can be handled as libcall here. */
+  { "libcall2",               0, 0, false, true, true, false,
+                              handle_libcall_attribute, NULL },
   { "sentinel",               0, 1, false, true, true, false,
 			      handle_sentinel_attribute, NULL },
   /* For internal use (marking of builtins) only.  The name contains space
@@ -5460,6 +5464,26 @@
 
   return NULL_TREE;
 }
+ 
+/* Handle a "libcall" attribute.  No special handling.  */
+
+static tree
+handle_libcall_attribute (tree *node, tree name,
+       tree ARG_UNUSED (args), int ARG_UNUSED (flags),
+       bool *no_add_attrs)
+{
+  if (TREE_CODE (*node) != FUNCTION_TYPE
+      && TREE_CODE (*node) != METHOD_TYPE
+      && TREE_CODE (*node) != FIELD_DECL
+      && TREE_CODE (*node) != TYPE_DECL)
+    {
+      warning (OPT_Wattributes,"%qs attribute only applies to functions",
+              IDENTIFIER_POINTER (name));
+      *no_add_attrs = true;
+    }
+
+   return NULL_TREE;
+}
 
 /* Handle a "sentinel" attribute.  */
 
