diff -ruN gcc-releases-gcc-13.4.0/gcc/cp/parser.cc gcc.work_13/gcc/cp/parser.cc
--- gcc-releases-gcc-13.4.0/gcc/cp/parser.cc	2025-06-05 18:03:58.000000000 +0200
+++ gcc.work_13/gcc/cp/parser.cc	2026-03-19 16:11:35.376401216 +0100
@@ -157,6 +157,7 @@
   RT_WHILE, /* while */
   RT_EXTERN, /* extern */
   RT_STATIC_ASSERT, /* static_assert */
+  RT_STATIC_WARNING, /* static_warning */
   RT_DECLTYPE, /* decltype */
   RT_OPERATOR, /* operator */
   RT_CLASS, /* class */
@@ -15207,7 +15208,7 @@
 	cp_lexer_consume_token (parser->lexer);
     }
   /* If the next token is `static_assert' we have a static assertion.  */
-  else if (token1->keyword == RID_STATIC_ASSERT)
+  else if (token1->keyword == RID_STATIC_ASSERT || token1->keyword == RID_STATIC_WARNING)
     cp_parser_static_assert (parser, /*member_p=*/false);
   /* If the next tokens after attributes is `using namespace', then we have
      a using-directive.  */
