diff -ruN gcc-releases-gcc-13.4.0/libgcc/configure ../_temp/gcc.work/libgcc/configure
--- gcc-releases-gcc-13.4.0/libgcc/configure	2025-06-05 18:03:58.000000000 +0200
+++ ../_temp/gcc.work/libgcc/configure	2026-03-17 23:02:21.259548559 +0100
@@ -5689,6 +5689,7 @@
 
 case $target_thread_file in
     aix)	thread_header=config/rs6000/gthr-aix.h ;;
+    amigaos)	thread_header=gthr-amigaos.h ;;
     dce)	thread_header=config/pa/gthr-dce.h ;;
     gcn)	thread_header=config/gcn/gthr-gcn.h ;;
     lynx)	thread_header=config/gthr-lynx.h ;;
