diff -ruN gcc-releases-gcc-13.4.0/libstdc++-v3/crossconfig.m4 ../_temp/gcc.work/libstdc++-v3/crossconfig.m4
--- gcc-releases-gcc-13.4.0/libstdc++-v3/crossconfig.m4	2025-06-05 18:03:58.000000000 +0200
+++ ../_temp/gcc.work/libstdc++-v3/crossconfig.m4	2026-03-18 00:10:02.059532218 +0100
@@ -299,6 +299,26 @@
 dnl # long_double_math_on_this_cpu in configure.ac, right after
 dnl # the expansion of the present macro.
     ;;
+  *-amigaos*)
+    AC_CHECK_HEADERS([nan.h ieeefp.h endian.h sys/isa_defs.h \
+      machine/endian.h machine/param.h sys/machine.h sys/types.h \
+      fp.h locale.h float.h inttypes.h])
+    SECTION_FLAGS='-ffunction-sections -fdata-sections'
+    AC_SUBST(SECTION_FLAGS)
+    GLIBCXX_CHECK_LINKER_FEATURES
+
+dnl # Although we are cross compiling for Amiga, we already built the
+dnl # first stage compiler: the compiler without any libraries such as
+dnl # libstdc++. This means that we DO have access to AC_CHECK_FUNCS
+dnl # as all the CRTs have been built and installed at the point of
+dnl # running the configure file for libstdc++-v3.
+dnl #
+dnl # Adding checks here:
+
+dnl #
+    AC_CHECK_FUNCS(acosf asinf atan2f atanf ceilf cosf coshf expf fabsf floorf fmodf frexpf sqrtf hypotf ldexpf log10f logf modff powf sinf sinhf tanf tanhf fabsl acosl asinl atanl atan2l ceill cosl coshl expl floorl fmodl frexpl sqrtl hypotl ldexpl logl log10l modfl powl sinl sinhl tanl tanhl)
+
+    ;;
   *)
     AC_MSG_ERROR([No support for this host/target combination.])
    ;;
