diff -rupN gcc.orig/gcc/analyzer/region-model-manager.cc gcc.work/gcc/analyzer/region-model-manager.cc
--- gcc.orig/gcc/analyzer/region-model-manager.cc	2025-12-08 09:27:36.990431912 +0100
+++ gcc.work/gcc/analyzer/region-model-manager.cc	2025-12-08 01:04:36.194446117 +0100
@@ -1119,7 +1119,7 @@ static void
 log_uniq_map (logger *logger, bool show_objs, const char *title,
 	      const hash_map<K, T*> &uniq_map)
 {
-  logger->log ("  # %s: %li", title, uniq_map.elements ());
+  logger->log ("  # %s: %u", title, uniq_map.elements ());
   if (!show_objs)
     return;
   auto_vec<const T *> vec_objs (uniq_map.elements ());
@@ -1143,7 +1143,7 @@ static void
 log_uniq_map (logger *logger, bool show_objs, const char *title,
 	      const consolidation_map<T> &map)
 {
-  logger->log ("  # %s: %li", title, map.elements ());
+  logger->log ("  # %s: %u", title, map.elements ());
   if (!show_objs)
     return;
