diff -rupN gcc.orig/gcc/config/rs6000/rs6000.c gcc.work/gcc/config/rs6000/rs6000.c
--- gcc.orig/gcc/config/rs6000/rs6000.c	2025-12-08 09:27:37.206431916 +0100
+++ gcc.work/gcc/config/rs6000/rs6000.c	2025-12-08 09:27:46.694432169 +0100
@@ -1373,6 +1373,12 @@ static const struct attribute_spec rs600
 #define TARGET_ASM_FUNCTION_PROLOGUE rs6000_output_function_prologue
 #undef TARGET_ASM_FUNCTION_EPILOGUE
 #define TARGET_ASM_FUNCTION_EPILOGUE rs6000_output_function_epilogue
+ 
+#ifdef TARGET_BASEREL
+extern void amigaos_function_end_prologue(FILE *);
+#undef TARGET_ASM_FUNCTION_END_PROLOGUE
+#define TARGET_ASM_FUNCTION_END_PROLOGUE amigaos_function_end_prologue
+#endif
 
 #undef TARGET_ASM_OUTPUT_ADDR_CONST_EXTRA
 #define TARGET_ASM_OUTPUT_ADDR_CONST_EXTRA rs6000_output_addr_const_extra
@@ -5886,6 +5892,7 @@ rs6000_file_start (void)
 #ifdef USING_ELFOS_H
   rs6000_machine = rs6000_machine_from_flags ();
   emit_asm_machine ();
+  /* AmigaOS: FIXME: The 'Disable .machine directive generation' is temporarily disabled. */
 #endif
 
   if (DEFAULT_ABI == ABI_ELFv2)
@@ -9005,6 +9012,14 @@ rs6000_legitimize_address (rtx x, rtx ol
       if (model != 0)
 	return rs6000_legitimize_tls_address (x, model);
     }
+ 
+#ifdef TARGET_BASEREL
+  if (TARGET_BASEREL
+      && amigaos_baserel_operand(x))
+    {
+      return amigaos_legitimize_baserel_address(x);
+    }
+#endif
 
   extra = 0;
   switch (mode)
@@ -10572,6 +10587,15 @@ rs6000_emit_move (rtx dest, rtx source,
 	}
       operands[1] = tmp;
     }
+ 
+#ifdef TARGET_BASEREL
+  if (/*GET_CODE (operands[1]) == SYMBOL_REF
+      && */TARGET_BASEREL
+      && amigaos_baserel_operand (operands[1]))
+    {
+       operands[1] = amigaos_legitimize_baserel_address (operands[1]);
+    }
+#endif
 
   /* 128-bit constant floating-point values on Darwin should really be loaded
      as two parts.  However, this premature splitting is a problem when DFmode
@@ -10832,6 +10856,15 @@ rs6000_emit_move (rtx dest, rtx source,
 					target, operands[1]));
 	      return;
 	    }
+ 
+#ifdef TARGET_BASEREL
+	  if (TARGET_BASEREL && amigaos_baserel_operand(operands[1]))
+	    {
+        emit_insn (gen_elf_base_high (target, gen_rtx_REG (Pmode, 2), operands[1]));
+        emit_insn (gen_elf_base_low (operands[0], target, operands[1]));
+        return;
+      }
+#endif
 
 	  emit_insn (gen_elf_high (target, operands[1]));
 	  emit_insn (gen_elf_low (operands[0], target, operands[1]));
@@ -14260,7 +14293,16 @@ print_operand_address (FILE *file, rtx x
 	   && CONSTANT_P (XEXP (x, 1)))
     {
       output_addr_const (file, XEXP (x, 1));
+#ifdef TARGET_BASEREL
+      if (TARGET_BASEREL && amigaos_baserel_operand(x))
+        {
+          fprintf (file, "@brel@l(%s)", reg_names[ REGNO (XEXP (x, 0)) ]);
+        }
+      else
+          fprintf (file, "@l(%s)", reg_names[ REGNO (XEXP (x, 0)) ]);
+#else
       fprintf (file, "@l(%s)", reg_names[ REGNO (XEXP (x, 0)) ]);
+#endif
     }
 #endif
   else if (toc_relative_expr_p (x, false, &tocrel_base_oac, &tocrel_offset_oac))
