diff -rupN gcc.orig/intl/dcigettext.c gcc.work/intl/dcigettext.c
--- gcc.orig/intl/dcigettext.c	2025-12-08 09:27:42.778432014 +0100
+++ gcc.work/intl/dcigettext.c	2025-12-08 09:27:46.378432156 +0100
@@ -148,7 +148,9 @@ extern int errno;
 char *getwd ();
 #  define getcwd(buf, max) getwd (buf)
 # else
+#   if !defined getcwd
 char *getcwd ();
+#   endif
 # endif
 # ifndef HAVE_STPCPY
 static char *stpcpy PARAMS ((char *dest, const char *src));
