diff -rupN gcc.orig/libcpp/line-map.c gcc.work/libcpp/line-map.c
--- gcc.orig/libcpp/line-map.c	2025-12-08 09:27:42.806432014 +0100
+++ gcc.work/libcpp/line-map.c	2025-12-08 09:27:46.378432156 +0100
@@ -1039,6 +1039,9 @@ linemap_ordinary_map_lookup (const line_
 
   const line_map_ordinary *cached = LINEMAPS_ORDINARY_MAP_AT (set, mn);
   /* We should get a segfault if no line_maps have been added yet.  */
+#ifdef __amigaos4__
+  linemap_assert(cached != 0);
+#endif
   if (line >= MAP_START_LOCATION (cached))
     {
       if (mn + 1 == mx || line < MAP_START_LOCATION (&cached[1]))
