diff -rupN gcc.orig/libiberty/configure.ac gcc.work/libiberty/configure.ac
--- gcc.orig/libiberty/configure.ac	2025-12-08 09:27:43.990432056 +0100
+++ gcc.work/libiberty/configure.ac	2025-12-08 09:27:46.722432170 +0100
@@ -618,6 +618,12 @@ if test -z "${setobjs}"; then
     ac_cv_func_getpagesize=yes
     ;;
 
+  *-*-amigaos*)
+    # current newlib lacks unsetenv(), so we custom versions
+    # of setenv() and unsetenv().
+    AC_LIBOBJ([setenv-amigaos])
+    ;;
+
   hppa*-*-hpux*)
     # Replace system snprintf and vsnprintf with libiberty implementations.
     AC_LIBOBJ([snprintf])
@@ -718,6 +724,7 @@ case "${host}" in
      *-*-mingw* | *-*-winnt*)	pexecute=pex-win32  ;;
      *-*-msdosdjgpp*)		pexecute=pex-djgpp  ;;
      *-*-msdos*)		pexecute=pex-msdos  ;;
+     *-*-amigaos*)		pexecute=pex-amigaos ;;
      *)				pexecute=pex-unix   ;;
 esac
 AC_SUBST(pexecute)
