diff -rupN gcc.orig/libssp/ssp.c gcc.work/libssp/ssp.c
--- gcc.orig/libssp/ssp.c	2025-12-08 09:27:44.202432064 +0100
+++ gcc.work/libssp/ssp.c	2025-12-08 09:27:47.110432186 +0100
@@ -58,6 +58,8 @@ see the files COPYING3 and COPYING.RUNTI
 #include <windows.h>
 #include <wincrypt.h>
 # define _PATH_TTY "CONOUT$"
+#elif defined (__amigaos4__)
+# define _PATH_TTY "CONSOLE:"
 #else
 # define _PATH_TTY "/dev/tty"
 #endif
@@ -89,7 +91,7 @@ __guard_setup (void)
         }
       CryptReleaseContext(hprovider, 0);
     }
-#else
+#elif !defined(__amigaos4__)
   int fd = open ("/dev/urandom", O_RDONLY);
   if (fd != -1)
     {
@@ -148,13 +150,13 @@ fail (const char *msg1, size_t msg1len,
         }
       close (fd);
     }
-
+#ifndef __amigaos4__
 #ifdef HAVE_SYSLOG_H
   /* Only send the error to syslog if there was no tty available.  */
   else
     syslog (LOG_CRIT, "%s", msg3);
 #endif /* HAVE_SYSLOG_H */
-
+#endif
   /* Try very hard to exit.  Note that signals may be blocked preventing
      the first two options from working.  The use of volatile is here to
      prevent optimizers from "knowing" that __builtin_trap is called first,
