diff -rupN gcc.orig/libstdc++-v3/include/std/mutex gcc.work/libstdc++-v3/include/std/mutex
--- gcc.orig/libstdc++-v3/include/std/mutex	2025-12-08 09:27:44.378432072 +0100
+++ gcc.work/libstdc++-v3/include/std/mutex	2025-12-08 09:27:47.602432207 +0100
@@ -780,8 +780,24 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
       once_flag::_Prepare_execution __exec(__callable);
 
       // XXX pthread_once does not reset the flag if an exception is thrown.
-      if (int __e = __gthread_once(&__once._M_once, &__once_proxy))
-	__throw_system_error(__e);
+      // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66146
+      // https://github.com/sba1/adtools/issues/82
+      int __e = 0;
+#if defined(__amigaos4__) && defined(__EXCEPTIONS)
+      try
+      {
+#endif
+        __e = __gthread_once(&__once._M_once, &__once_proxy);
+#if defined(__amigaos4__) && defined(__EXCEPTIONS)
+      }
+      catch (...)
+      {
+        __gthread_once_unlock(&__once._M_once);
+        throw;
+      }
+#endif
+      if (__e)
+        __throw_system_error(__e);
     }
 
 #else // _GLIBCXX_HAS_GTHREADS
